src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.CodeFixes/ApiControllerCodeFixer.cs

```diff
@@ -0,0 +1,100 @@
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+            if (root is null)
+            {
+                return;
+            }
+
+            var node = root.FindNode(context.Span, false, true);
+
+            if (node is null)
+            {
+                return;
+            }
+
+            // Register a code action that will invoke the fix.
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    CodeFixResources.ApiControllerTitle,
+                    cancellationToken => ReplaceBaseClass(context.Document, node, cancellationToken),
+                    nameof(CodeFixResources.ApiControllerTitle)),
+                context.Diagnostics);
+        }
+
+        /// <summary>
+        /// Built with the assumption that the <paramref name="node"/> is a IdentifierNameSyntax or QualifiedNameSyntax that indicates an ApiController BaseType.
+        /// </summary>
+        /// <param name="document">The document that contains a class to fix.</param>
+        /// <param name="node">A class that is ApiController.</param>
+        /// <param name="cancellationToken">May request cancellation.</param>
+        /// <returns>A document where at least one ApiController was removed.</returns>
+        private async Task<Document> ReplaceBaseClass(Document document, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var mvcBaseType = generator.QualifiedName(
+                CreateMvcNamespace(generator),
+                generator.IdentifierName(GoodClassName))
+                .WithAdditionalAnnotations(Simplifier.Annotation, Simplifier.AddImportsAnnotation);
+
+            var baseType = GetBaseType(node, generator);
+
+            var newRoot = generator.ReplaceNode(root, baseType, mvcBaseType);
+
+            var newDocument = document.WithSyntaxRoot(newRoot);
+
+            newDocument = await ImportAdder.AddImportsAsync(newDocument, Simplifier.AddImportsAnnotation, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            return newDocument;
+        }
+
+        private static SyntaxNode GetBaseType(SyntaxNode node, SyntaxGenerator generator)
+        {
+            var types = generator.GetBaseAndInterfaceTypes(node);
+            if (types == null || types.Count == 0)
+            {
+                types = generator.GetBaseAndInterfaceTypes(node.Parent);
+            }
+
+            return types[0];
+        }
+
+        private static SyntaxNode CreateMvcNamespace(SyntaxGenerator generator) => generator.QualifiedName(CreateAspNetCoreNamesapce(generator), generator.IdentifierName("Mvc"));
```

Comment:
Could these be generalized? So we could pass in a fully qualified name and it spits out the syntax node we want?

I could foresee this being used to replace additional base types

Reply:
Yeah, I'm going to replace this with a QualifiedName helper that will enable us to build the syntax from string. I think we could evolve that going forward to do some other QualifiedName analysis